TW-141 Activities are idempotent (#1767)
temporalio/documentation: docs-src/app-dev-context/developing-activities.md

---

One of the primary things that Workflows do is orchestrate the execution of Activities.

> An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as making a calculation, calling a third-party API, or transcoding a media file.

Reviewer: ```suggestion
An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as querying a database, calling a third-party API, or transcoding a media file.
```

> Activities are normal function/method executions, each one intended to execute a single, well-defined action (either short or long-running) such as making a calculation, calling a third party API, or transcoding a media file.
> Activities can interact with world outside the Temporal Platform or use a Temporal Client to interact with a Cluster.

Reviewer 2: ```suggestion
An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as making a calculation, calling a third-party API, or transcoding a media file.
An Activity can interact with world outside the Temporal Platform or use a Temporal Client to interact with a Cluster.
```

One of the primary things that Workflows do is orchestrate the execution of Activities.
Activities are normal function/method executions that can interact with the world.
For the Workflow to be able to execute the Activity, we must define the [Activity Definition](/concepts/what-is-an-activity-definition).

> Temporal requires that Activities are idempotent. Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.

Reviewer: Like to the Idempotency section in the concept?

Reviewer 3: I'll point out that this wording says that Temporal _requires_ Activities to be idempotent, but the wording in `docs-src/concepts/what-is-an-activity.md` says that Temporal _recommends_ them to be idempotent. 

Personally, I think that recommends (or perhaps "strongly recommends") is better wording, from the perspective that Temporal can do its job whether or not the Activity is idempotent (in contrast to whether or not a Workflow is deterministic, a violation of which can lead to failure). Others have tried to make the case that "required" is better. 

I will leave it up to the Docs team to decide which is more accurate, but I do suggest that Activity idempotence will confuse people if it's described as "recommended" in one place and "required" in another.

Reviewer 4: It doesn't require, and there are times when idempotency isn't possible / supported by a downstream (possibly 3rd party) API

Reviewer 4: 🤷‍♂️ I guess "in order for your application to be correct / always function as intended in the presence of failure, activities must be idempotent" can be shortened to "activities must be idempotent"

> Temporal recommends that Activities are idempotent. Activities are idempotent if multiple applications of that operation do not change the state of the system beyond the initial application.

Reviewer 5: More than recommends. They must be idempotent or things can go wrong.
